core/trino-main/src/main/java/io/trino/sql/rewrite/ShowQueriesRewrite.java

```diff
@@ -389,16 +398,46 @@ protected Node visitShowColumns(ShowColumns showColumns, Void context)
             if (!metadata.schemaExists(session, new CatalogSchemaName(tableName.getCatalogName(), tableName.getSchemaName()))) {
                 throw semanticException(SCHEMA_NOT_FOUND, showColumns, "Schema '%s' does not exist", tableName.getSchemaName());
             }
-            if (metadata.getView(session, tableName).isEmpty() &&
-                    metadata.getTableHandle(session, tableName).isEmpty()) {
+
+            Optional<TableHandle> tableHandle = metadata.getTableHandle(session, tableName);
+            if (metadata.getView(session, tableName).isEmpty() && tableHandle.isEmpty()) {
                 throw semanticException(TABLE_NOT_FOUND, showColumns, "Table '%s' does not exist", tableName);
             }
 
-            accessControl.checkCanShowColumns(session.toSecurityContext(), tableName.asCatalogSchemaTableName());
-
+            Relation relation = from(tableName.getCatalogName(), COLUMNS.getSchemaTableName());
             Expression predicate = logicalAnd(
                     equal(identifier("table_schema"), new StringLiteral(tableName.getSchemaName())),
                     equal(identifier("table_name"), new StringLiteral(tableName.getObjectName())));
+            if (tableHandle.isPresent()) {
+                // Queries against information_schema should never fail. In case of an issue of getting
+                // table metadata the error could be swallowed by using information_shhema.
+                TableMetadata tableMetadata = metadata.getTableMetadata(session, tableHandle.orElseThrow());
+
+                ImmutableList.Builder<Expression> rows = ImmutableList.builder();
+                for (int i = 0; i < tableMetadata.getColumns().size(); i++) {
+                    ColumnMetadata columnMetadata = tableMetadata.getColumns().get(i);
+                    rows.add(row(
+                            cast(toExpression(i + 1), VARCHAR),
```

Comment:
not a varchar, and casting to varchar  will result in wrong ordering when table has 10 columns